lua/lazyvim/plugins/ui.lua

```diff
@@ -29,6 +29,10 @@ return {
         end)
       end
     end,
+    config = function(_, opts)
+      require("notify").setup(opts)
```

Comment:
Essentially I tested that in the first place, and custom `opts` are not applied.

Reply:
Excuse my ignorance, but how do you test about custom `opts` being applied?

Reply:
Hi @dpetka2001,

The opts step is only for specifying and merging opts, as far as I know.
The user can specify config = true, and lazy.nvim will try to require the plugin and call its setup method with the provided opts, "magically".
When the user specifies the config function however, the require and setup has to be written explicitly.

Reply:
Yes i understand that. My question is about why in this specific situation, we need the config function. Per the docs
> opts should be a table (will be merged with parent specs), return a table (replaces parent specs) or should change a table. The table will be passed to the Plugin.config() function. Setting this value will imply Plugin.config()

Because i have seen in quite a few places configuring only `opts` without the config function, even doing `opts ={} ` instead of `config = true`, which has been mentioned to be almost equivalent. It seems that changing the values for `opts` in this case is not being respected unless we pass them through the config function. Read about how i tried with the `timeout` value in my previous post and if you like try it out yourself and share your experience.

Reply:
My previous comment was triggered by replaying your steps. I hope I understand them, my apologies if I don't. 

You wrote:
> I don't believe this is needed?? Since opts by default will call the plugin's setup function? Can you test without this line 33 to see if it works?

This is not correct. Opts does not call the plugin's setup function. Plugin.config() does,
as is stated in your `opts` citation

Citation from docs, `cond` keyword:
>config is executed when the plugin loads. The default implementation will automatically run require(MAIN).setup(opts). Lazy uses several heuristics to determine the plugin's MAIN module automatically based on the plugin's name. See also opts. To use the default implementation without opts set config to true.

There is a default config implementation doing the work. As soon as the user enters a custom implementation, the default implementation is gone. The user has to do the work in the new function, passing the merged options to the plugin. If the merged options are not passed to the setup of ie. nvim-notify, the defaults of that plugin will apply.

Example, *stock lazyvim starter*, only adding new file lua/plugins/notify.lua
```lua
return {
  {
    "rcarriga/nvim-notify",
    opts = {
      timeout = 15000,
    },
  },
}
```
Result: It takes a long time for the notification to disappear.

Example, add config = true
```lua
return {
  {
    "rcarriga/nvim-notify",
    opts = {
      timeout = 15000,
    },
    config = true,
  },
}
```
Result: The same, opts.timeout is processed correctly, by the default config function.

Example, with a function doing nothing
```lua
return {
  {
    "rcarriga/nvim-notify",
    opts = {
      timeout = 15000,
    },
    config = function(_, opts) end,
  },
}
```
Result: Nvim-notify's setup was not invoked with opts. The default of nvim-notify is 5s.

Reply:
Ohhhh ok I see. I had understood that `opts` is a table that gets passed to `Plugin.config()`. No problem there, I just didn't express myself correctly by saying it calls the plugin's setup function.. But because this PR changes the `config` function to require load the notify telescope extension, we have to also pass the `opts` now, since the default behavior of the config function is being changed by the user? That's why the `opts` don't take effect if we don't pass them to the config function.
Thank you for your informative explanation @abeldekat . Really appreciate it :smile:

Reply:
You are most welcome!

> But because this PR changes the config function to require load the notify telescope extension, we have to also pass the opts now, since the default behavior of the config function is being changed by the user?

I would say that this PR replaced the default, implied, config function. Now, the new function has to do the "implied" work, in addition to setting up telescope for nvim-notify.

Adding to the review:
I do not think requiring telescope in the new code is a good idea. Now, when telescope is disabled, nvim-notify will crash.

Reply:
Yes, I believe a better place to load the extension would be in the `telescope` spec or in this spec just specify a keymap instead like
```lua
      {
        "<leader>unt",
        function()
          require("telescope").extensions.notify.notify()
        end,
        desc = "Open Notifications (Telescope)",
      },
```

Reply:
I think special care is needed when plugins "hard require" other plugins.
Nvim-notify is loaded when noice is loaded, on the VeryLazy event. Furhtermore, in my opinion, nvim-notify is not a telescope-only plugin, it just provides a convenient integration.
Telescope, one of the 'bigger' plugins,  is only loaded on command or custom keys. 

Regarding the approach in this pull request:
Requiring telescope in nvim-notify's config function would effectively load telescope on VeryLazy as well, impacting performance.

The keymap you mentioned makes sense, however: What happens when the user disables nvim-notify?

May I ask you to have a look at this [pull request](https://github.com/LazyVim/LazyVim/pull/1076)? Folke solved a similar coupling issue.

Reply:
The keymap would be under the `keys` section of `nvim-notify` spec. If the user disables nvim-notify, then no keymap will be available.

Reply:
@dpetka2001 
> The keymap would be under the keys section of nvim-notify spec. If the user disables nvim-notify, then no keymap will be available. That's where I have it. Not in telescope spec.

But, when a user disables *telescope*, that keymap on nvim-notify will still be available and crash on invocation. It's a chicken and egg problem.

For an individual user it is much easier to activate the telescope addon of nvim-notify.
In LazyVim's base setup, the consequences of a user disabling noice, nvim-notify or telescope need to be taken into account.

I created a config to test and discuss, using *stock lazyvim starter*, only adding file lua/plugins/notify.lua:
```lua
--[[
A candidate for LazyVim's lua/plugins/ui.lua
--]]

local Util = require("lazyvim.util")

return {
  {
    "telescope.nvim",
    optional = true,
    keys = function(_, keys)
      --
      -- HACK for now:
      -- this condition returns true even when nvim-notify is disabled
      -- Cause: In lazy.nvim fix-disabled is called **after** calculate-handlers.
      -- @See lazy.nvim, plugins.lua, the parse method
      if Util.has("nvim-notify") then
        vim.list_extend(keys, {
          {
            "<leader>sN", -- "n" is assigned to noice
            function()
              -- do not crash when nvim-notify is disabled, see HACK...
              if Util.has("nvim-notify") then
                require("telescope").extensions.notify.notify()
              end
            end,
            desc = "Open Notifications (Telescope)",
          },
        })
      end
    end,
  },

  {
    "rcarriga/nvim-notify",
    config = function(_, opts)
      require("notify").setup(opts)
      if Util.has("telescope.nvim") then
        require("lazyvim.util").on_load("telescope.nvim", function()
          -- when telescope is loaded, the extension will also be loaded.
          require("telescope").load_extension("notify")
        end)
      end
    end,
  },

  -- TESTING
  -- disable noice and nvim-notify
  -- { "noice.nvim", enabled = false },
  -- { "nvim-notify", enabled = false },

  -- disable telescope
  -- { "telescope.nvim", enabled = false },
  -- END TESTING
}
```

Reply:
In addition to my previous [comment](https://github.com/LazyVim/LazyVim/pull/1306#discussion_r1289108126), the code below is also possible:

```lua
--[[
A candidate for LazyVim's lua/plugins/ui.lua
--]]

-- stylua: ignore start
-- if true then return {} end
-- stylua: ignore end

local Util = require("lazyvim.util")

return {

  {
    "rcarriga/nvim-notify",
    keys = function(_, keys)
      --
      -- HACK:
      -- this condition returns true even when telescope.nvim is disabled
      -- Cause: In lazy.nvim fix-disabled is called before calculate-handlers.
      -- See lazy.nvim, plugins.lua, the parse method:
      --
      if Util.has("telescope.nvim") then
        vim.list_extend(keys, {
          {
            "<leader>sN", -- "n" is assigned to noice
            function()
              -- do not crash when telescope.nvim is disabled, see HACK...
              if Util.has("telescope.nvim") then
                require("telescope").extensions.notify.notify()
              end
            end,
            desc = "Open Notifications (Telescope)",
          },
        })
      end
    end,
    config = function(_, opts)
      require("notify").setup(opts)
      if Util.has("telescope.nvim") then
        require("lazyvim.util").on_load("telescope.nvim", function()
          -- when telescope is loaded, the extension will also be loaded.
          require("telescope").load_extension("notify")
        end)
      end
    end,
  },

  -- TESTING
  -- disable noice and nvim-notify
  -- { "noice.nvim", enabled = false },
  -- { "nvim-notify", enabled = false },

  -- disable telescope
  -- { "telescope.nvim", enabled = false },
  -- END TESTING
}

```

Reply:
I have just the keymap in notify's `keys` wrapped into a `Util.has("telescope.nvim")` like you also suggested. I disabled Telelescope and notify doesn't crash. I used the following code for testing
```lua
      {
        "<leader>unt",
        function()
          if Util.has("telescope.nvim") then
            require("telescope").extensions.notify.notify({
              initial_mode = "normal",
              -- layout_strategy = "vertical",
            })
          else
            require("which-key").register({
              ["<leader>unt"] = "which_key_ignore",
            })
          end
        end,
        desc = "Open Notifications (Telescope)",
      },
```
I don't believe there is any need to load the notify extension in Telescope config function since we go the keymap way. The only caveat with this approach is that the keymap still registers under `which-key` until you invoke the function with the keybinding and then it gets removed from being registered in which-key.

Reply:
Btw I didn't know about the `on_load` function. Thank you for giving me the opportunity for learning something new :smile:

Reply:
The on_load function was added in this [pull-request](https://github.com/LazyVim/LazyVim/pull/1076?)
Thank you as well! I am learning a lot by engaging in this conversation.

I am surprised that there is no need to load the notify extension. The requirement is specified as a note [here](https://github.com/rcarriga/nvim-notify#viewing-history).

Could you paste your config for me to try?

Reply:
```lua
  {
    "rcarriga/nvim-notify",
    keys = {
      {
        "<leader>und",
        function()
          require("notify").dismiss({ silent = true, pending = true })
        end,
        desc = "Delete all Notifications",
      },
      {
        "<leader>unt",
        function()
          if Util.has("telescope.nvim") then
            require("telescope").extensions.notify.notify({
              initial_mode = "normal",
              -- layout_strategy = "vertical",
            })
          else
            require("which-key").register({
              ["<leader>unt"] = "which_key_ignore",
            })
          end
        end,
        desc = "Open Notifications (Telescope)",
      },
    },
  },
```
I have only this without loading the notify extension in telescope config and telescope is not loaded until I press some telescope keybinding.

Reply:
Thank you both for this conversation! :heart:

Reply:
The solution @dpetka2001 describes works perfectly.

I *really* do not understand why the notify extension does not have to be loaded. It's in the docs of nvim-notify. Any ideas?
As a side-effect, it's nice no config function is needed for nvim-notify.

I would suggest to remove the hard require on which-key. I also removed the key entirely, thus the function is only entered once in case telescope is disabled.

Currently, in case telescope is disabled, the initial presence of the key cannot be avoided. See the [hack](https://github.com/LazyVim/LazyVim/pull/1306#discussion_r1289116960) in my comment. 
The `keys` *keyword*, allowing a function instead of a list,  is applied before lazy.nvim removes disabled plugins from its active list.


```lua
{
        "<leader>unt",
        function()
          if Util.has("telescope.nvim") then
            require("telescope").extensions.notify.notify()
          else
            Util.on_load("which-key.nvim", function()
              require("which-key").register({
                ["<leader>unt"] = "which_key_ignore",
              })
              vim.keymap.del("n", "<leader>unt")
            end)
          end
        end,
        desc = "Open Notifications (Telescope)",
}
```

Reply:
Plz check folke's answer in this [comment](https://github.com/LazyVim/LazyVim/issues/283#issuecomment-1433633320). It partially explains why there is no need for the notify extension(it's the same concept I believe). I like that you removed the hard require on which-key and see the `on_load` in effect. I haven't looked into `lazy.nvim`'s code that much (maybe it's time I should? :stuck_out_tongue: ), so I'm really grateful about you sharing your knowledge.

Reply:
Thanks for pointing me to that comment! Again, I would like to return the compliment. I think you do a great job in assisting this project!

Reply:
@dpetka2001 @abeldekat Thank you very much for the constructive suggestions. https://github.com/LazyVim/LazyVim/pull/1306#discussion_r1289230224 added a new keymap that takes various scenarios into consideration. However, I have two concerns/questions:
1. How to also enable `:Telescope notify`? Is the solution in https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/util/project.lua adoptable?
2. Is it over-engineered? I searched telescope in this project but didn't find any single place to check its presence...

Reply:
Hi @MoetaYuko,

Yes, I also think it is over-engineered. Too many guards are needed. The keymap is always created, whereas a condition would make sense.

The challenge is unique though, given that I think we should care about the lazy loading and the different combinations of actually active plugins.
The big difference with the "project.nvim" solution: That plugin is *only* a telescope dependency: There are no other specs for "project.nvim". 

In my opinion, this might be a perfect use case for the change I hope to propose. It is discussed [here](https://github.com/folke/lazy.nvim/discussions/949#discussioncomment-6598353)

Reply:
If you would like to see how this can work without the guards, please follow the following steps.


- Step one: Local lazy.nvim:
```
mkdir ~/projects
git clone https://github.com/abeldekat/lazy.nvim.git ~/projects/lazy.nvim
cd ~/projects/lazy.nvim
git checkout discard_deps_revisited
```
- Step two: Lazytest with modified starter including nvim-notify config:
```
git clone https://github.com/abeldekat/starter.git ~/.config/lazytest
cd ~/.config/lazytest
git checkout discard_deps_revisited
```
- Run lazytest
```
cd ~/.config/lazytest
LAZY=~/projects/lazy.nvim NVIM_APPNAME=lazytest nvim
```
- Quit when all is installed and restart lazytest
- Inspect `lua/plugins/notify.lua`, test with different scenario's, see its TESTING section.

If all is well, the keymap will only show when all plugins are active.

Reply:
@MoetaYuko Currently if we choose to go with the keymap way, then `:Telescope notify` will not be available the first time. You will first have to invoke it with keybinding and afterwards it will also be available as a command. For the command to also be available, we would have to go the `load_extension` way for notify in Telescope. Of course we could go both ways, but the `load_extension` should happen in Telescope config function and not in notify's spec as that would impact Telescope's lazy loading.

@abeldekat I tried your proposed changes and indeed does work exactly like you described with your branch of `lazy.nvim`. I'm sorry, but I cannot help any further as I don't have programming background and my Lua knowledge is just enough to tinker around the Lazyvim ecosystem which I have accustomed to. I can only say it works exactly like you described without having the knowledge to judge if there could be any ramifications by the changes you want to propose. One minor note, you might want to change the `<leader>un` which-key namespace, because it says "Dismiss all notifications" and I was looking around to see where the "Open telescope notifications" keybinding was (decided to just press the combination in the config you provided and it worked as expected). As always, thank you for your contributions and I wish you good luck with your `lazy.nvim` PR.

Reply:
@dpetka2001 ,

Thanks for the kind answer! You have just helped me significantly by confirming the solution works. The proposal is from @mehalter. I searched the issues, and found [one](https://github.com/folke/lazy.nvim/issues/868) expressing the idea.

Reply:
Yes, the key is not added when telescope is disabled. It definitely fixes the caveat with the keybinding solution that we had in a previous post. So, if I understand correctly, your proposed solution will not parse dependencies if a plugin is disabled, whereas as things are now, dependencies will get parsed even if a plugin is disabled?

Reply:
> So, if I understand correctly, your proposed solution will not parse dependencies if a plugin is disabled, whereas as things are now, dependencies will get parsed even if a plugin is disabled?

Correct!

Reply:
@MoetaYuko 

> How to also enable :Telescope notify? 

Using the modified [lazy.nvim](https://github.com/LazyVim/LazyVim/pull/1306#discussion_r1289746423), it can be done by adding the following snippet to ui.lua:
```
{
    "telescope.nvim", -- "contribute" to telescope defined in editor.lua
    opts = function() -- use opts, so the user can define a config function
      if Util.has("nvim-notify") then -- the user can disable notify
        require("telescope").load_extension("notify")
      end
    end,
    dependencies = {
      {
        "nvim-notify", --"contribute" to nvim-notify defined in ui.lua
        keys = {
          {
            "<leader>sN",
            function()
              require("telescope").extensions.notify.notify()
            end,
            desc = "Search Notifications",
          },
        },
      },
    },
  }
```

@dpetka2001 
I think the shortcut "leader s N" is also worth considering. The ui menu allows for all notifications to be discarded, a "UI" action. The search menu caters for searching items. I noticed noice uses "leader s n"

I uploaded the new code to the modified starter.

Reply:
@jyuan0,

You showed interest in this conversation and I read your awesome answer in this [issue](https://github.com/LazyVim/LazyVim/issues/1316#issuecomment-1672683656). If you have the time I would be very interested in your thoughts.

Reply:
@abeldekat I totally agree that handling disabled deps should be implemented in lazy.nvim rather than us adding guards everywhere, and your above snippet looks neat. Maybe getting your lazy.nvim changes upstreamed before we update this PR?

Reply:
@abeldekat I don't have any problem with what keybinding you will choose. I just mentioned it in my previous post because `<leader>un` was not a which-key namespace in which you usually expect to find further keys and I stumbled a bit searching for the actual keybinding that you had provided in your post at the time. That's all. Since `Noice history` is also under the search namespace, I agree that `Notify history` would belong there as well.

Reply:
@abeldekat I would appreciate it if you could take the time to explain why you chose to put `load_extension` in the Telescope `opts` instead of `config`. I usually have
```lua
    config = function(_, opts)
      local telescope = require("telescope")
      telescope.setup(opts)
      telescope.load_extension("project")
    end,
```
to ensure that first Telescope gets set up with the `opts` that are passed to it and then load the extensions. Some extensions, like `telescope-project` are dependent on Telescope `opts` where they can be defined the options. See my dotfiles [here](https://github.com/dpetka2001/dotfiles/blob/8566fe1f4609398c22b02a2004d648b69246758b/dot_config/nvim/lua/plugins/telescope.lua#L106-L115). I understand that extensions like nvim-notify, which don't get configured in Telescope `opts.extensions`, can be configured to load when they are called. I just keep this under `config` for good practice and so that I don't do any mistake by accident. Could you explain the difference plz when you have time?

Reply:
@dpetka2001

This indeed requires explaining. Note that my reasoning is by for not authoritative! I have little in depth knowledge regarding lazy.nvim's code in this processing phase.

You gave me very important information, some of that I already knew: See [here](https://github.com/LazyVim/LazyVim/issues/283#issuecomment-1433633320)

I assume LazyVim needs to apply much more consideration in configuring a base for plugins than the end user. Thus, your config makes sense to me.

Also, *unofficialy*, I think of all individual spec fragments resulting in a concrete  plugin definition as being "contributions".

As far as I know, the final spec of a certain plugin has only *one* `config` function. The keywords `opts, ft, event, keys` used by individual contributions are merged/extended.

I noticed that telescope in `lua/plugins/editor.lua` does not have a config function, implying:
1. The user could have defined one, in which case the new config function in LazyVim is overridden by the user's config function
2. The user has the freedom to define one...

Thus, I chose to use the opts function to load the telescope extension. This usage is not complying to the docs:
>opts should be a table (will be merged with parent specs), return a table (replaces parent specs) or **should change a table**. The table will be passed to the Plugin.config() function. Setting this value will imply Plugin.config()

However, when the purpose is to keep the config function available to the user( without the user having to overwrite any existing config function), I think it makes sense.

Reply:
I see. It makes sense as you put it. Thanks for your detailed explanation.

Reply:
@abeldekat,

> @jyuan0,
>
> You showed interest in this conversation and I read your awesome answer in this https://github.com/LazyVim/LazyVim/issues/1316#issuecomment-1672683656. If you have the time I would be very interested in your thoughts.

Thanks, I was inspired by this awesome thread :smiley:.

I don't think I really have anything meaningful to add. The three of you have already **thoroughly** considered the possible implementations. I really appreciate the thought you have all put into lazy-loading, enabled/disabled plugins, and how to best implement things.

Regarding the proposed implementations: I think it makes sense to wait to see if the proposed lazy.nvim changes get merged upstream. This change:
```lua
{
    "telescope.nvim",
    opts = function()
      if Util.has("nvim-notify") then
        require("telescope").load_extension("notify")
      end
    end,
    dependencies = {
      {
        "nvim-notify",
        keys = {
          {
            "<leader>sN",
            function()
              require("telescope").extensions.notify.notify()
            end,
            desc = "Search Notifications",
          },
        },
      },
    },
  }
```
looks preferable to this:
```lua
{
        "<leader>unt",
        function()
          if Util.has("telescope.nvim") then
            require("telescope").extensions.notify.notify()
          else
            Util.on_load("which-key.nvim", function()
              require("which-key").register({
                ["<leader>unt"] = "which_key_ignore",
              })
              vim.keymap.del("n", "<leader>unt")
            end)
          end
        end,
        desc = "Open Notifications (Telescope)",
}
```
I am wondering if it's necessary to register the key with which-key before deleting it, but that's minor and might not even matter if the proposed lazy.nvim changes get upstreamed.

Again, thanks for this awesome conversation and PR review!

Reply:
I created the [PR](https://github.com/folke/lazy.nvim/pull/1000) for `lazy.nvim`, proposing the changes we discussed.

Reply:
Good luck with your PR :smile:

Reply:
@folke improved the PR.
The changes are merged into lazy.nvim v10.4.0!